### ISSUE TITLE
client (v7 v7.6.0) crashing database while inserting record on complex secondary indexed column

### ISSUE THREAD

**Issue by asado** [aerospike/aerospike-client-go#448]
secondary Index:
```
	indexTask, aeroErr := s.aeroDB.CreateComplexIndex(
		aero.NewWritePolicy(0, 0),
		"test",
		"set_name",
		"tokens_key_index",
		"sotBin",
		aero.STRING,
		aero.ICT_MAPVALUES,
		aero.CtxMapKey(aero.NewWildCardValue()), // Dynamic columnID key
		aero.CtxMapKey(aero.StringValue("T")), 
	)
```
The above secondary index has a mistake of using CtxMapKey instead of CtxMapValue but it should not crash the database while inserting the below.
```
{
    "foo": {
       "T": {
          "cat1": "bar1",
          "cat2": "bar2"
       },
      "A": {
          "abc": "xyz"
      }
    }
}
```

**Comment by khaf** [comment]
I can't reproduce this issue on server v7.1. Do you have a gist that shows the issue? Here's my code:

```go
package main

import (
	"bytes"
	"flag"
	"log"
	"os"
	"time"

	aero "github.com/aerospike/aerospike-client-go/v7"
)

var (
	Host = flag.String("h", "localhost", "Aerospike server seed hostnames or IP addresses.")
	Port = flag.Int("p", 3000, "Aerospike server seed hostname or IP address port number.")
)

func main() {
	flag.Parse()

	var buf bytes.Buffer
	logger := log.New(&buf, "", log.LstdFlags|log.Lshortfile)
	logger.SetOutput(os.Stdout)

	// connect to the host
	clientPolicy := aero.NewClientPolicy()

	h := []*aero.Host{
		aero.NewHost(*Host, *Port),
	}
	log.Println("Hosts:", h)
	client, err := aero.NewClientWithPolicyAndHost(clientPolicy, h...)
	if err != nil {
		log.Fatalln(err)
	}

	if err = client.DropIndex(nil, "test", "test", "tokens_key_index"); err != nil {
		log.Fatalln(err)
	}
	indexTask, err := client.CreateComplexIndex(
		aero.NewWritePolicy(0, 0),
		"test",
		"test",
		"tokens_key_index",
		"sotBin",
		aero.STRING,
		aero.ICT_MAPVALUES,
		// aero.CtxMapValue(aero.NewWildCardValue()), // Dynamic columnID key
		aero.CtxMapKey(aero.NewWildCardValue()), // Wrong, crashes the DB
		aero.CtxMapKey(aero.StringValue("T")),
	)

	if err != nil {
		log.Fatalln(err)
	}

	if err = <-indexTask.OnComplete(); err != nil {
		log.Fatalln(err)
	}

	time.Sleep(time.Second * 3)

	bins := aero.BinMap{
		"foo": map[string]any{
			"T": map[string]any{
				"cat1": "bar1",
				"cat2": "bar2",
			},
			"A": map[string]any{
				"abc": "xyz",
			},
		},
	}

	key, _ := aero.NewKey("test", "test", 1)
	if err = client.Put(nil, key, bins); err != nil {
		log.Fatalln(err)
	}

	stmt := aero.NewStatement("test", "test")
	if err = stmt.SetFilter(aero.NewContainsFilter("foo", aero.ICT_MAPVALUES, "foo",
		aero.CtxMapValue(aero.NewWildCardValue()), // Dynamic columnID key
		aero.CtxMapKey(aero.StringValue("T")),
	)); err != nil {
		log.Fatalln(err)
	}
	rs, err := client.Query(nil, stmt)
	if err != nil {
		log.Fatalln(err)
	}

	for res := range rs.Results() {
		if res.Err != nil {
			log.Fatalln(err)
		}

		log.Println(res.Record.Bins)
	}

	log.Println("Done!")
}
```

**Comment by khaf** [comment]
Note that you don't have a `sotBin` anywhere in the sample data.

**Comment by asado** [comment]
```
{
    "foo": {
       "T": {
          "cat1": "bar1",
          "cat2": "bar2"
       },
      "A": {
          "abc": "xyz"
      }
    }
}
```
The above is content of sotBin. Below code will crash:

``` golang
package main

import (
	"bytes"
	"flag"
	"log"
	"os"
	"time"

	aero "github.com/aerospike/aerospike-client-go/v7"
)

var (
	Host = flag.String("h", "localhost", "Aerospike server seed hostnames or IP addresses.")
	Port = flag.Int("p", 3000, "Aerospike server seed hostname or IP address port number.")
)

func main() {
	flag.Parse()

	var buf bytes.Buffer
	logger := log.New(&buf, "", log.LstdFlags|log.Lshortfile)
	logger.SetOutput(os.Stdout)

	// connect to the host
	clientPolicy := aero.NewClientPolicy()

	h := []*aero.Host{
		aero.NewHost(*Host, *Port),
	}
	log.Println("Hosts:", h)
	client, err := aero.NewClientWithPolicyAndHost(clientPolicy, h...)
	if err != nil {
		log.Fatalln(err)
	}

	defer client.Close()

	if err = client.DropIndex(nil, "test", "test", "tokens_key_index"); err != nil {
		log.Fatalln(err)
	}
	indexTask, err := client.CreateComplexIndex(
		aero.NewWritePolicy(0, 0),
		"test",
		"test",
		"tokens_key_index",
		"sotBin",
		aero.STRING,
		aero.ICT_MAPVALUES,
		// aero.CtxMapValue(aero.NewWildCardValue()), // Dynamic columnID key
		aero.CtxMapKey(aero.NewWildCardValue()), // Wrong, crashes the DB
		aero.CtxMapKey(aero.StringValue("T")),
	)

	if err != nil {
		log.Fatalln(err)
	}

	if err = <-indexTask.OnComplete(); err != nil {
		log.Fatalln(err)
	}

	time.Sleep(time.Second * 3)

	bins := aero.BinMap{
		"sotBin": map[string]any{
			"T": map[string]any{
				"cat1": "bar1",
				"cat2": "bar2",
			},
			"A": map[string]any{
				"abc": "xyz",
			},
		},
	}

	key, _ := aero.NewKey("test", "test", 1)
	if err = client.Put(nil, key, bins); err != nil {
		log.Fatalln(err)
	}

	stmt := aero.NewStatement("test", "test")
	if err = stmt.SetFilter(aero.NewContainsFilter("foo", aero.ICT_MAPVALUES, "foo",
		aero.CtxMapValue(aero.NewWildCardValue()), // Dynamic columnID key
		aero.CtxMapKey(aero.StringValue("T")),
	)); err != nil {
		log.Fatalln(err)
	}
	rs, err := client.Query(nil, stmt)
	if err != nil {
		log.Fatalln(err)
	}

	for res := range rs.Results() {
		if res.Err != nil {
			log.Fatalln(err)
		}

		log.Println(res.Record.Bins)
	}

	log.Println("Done!")
}
```

**Comment by khaf** [comment]
This does not crash the server v7.1.0.0CE for me. Could you please share your server version, edition and config file (with important info redacted)?

**Comment by asado** [comment]
I am using aerospike image `aerospike/aerospike-server:7.0.0.0_2` with default config.

docker-compose:
```
  aerospike-unit-db:
    container_name: aerospike-unit-db
    image: aerospike/aerospike-server:7.0.0.0_2
    ports:
      - "3000:3000"
      - "3001:3001"
      - "3002:3002"
      - "3003:3003"
 ```
Using the above code, it crashes with below logs:
```
ug 27 2024 11:39:20 GMT: INFO (info): (thr_info.c:3039) sindex-create: request received for test:tokens_key_index via info
Aug 27 2024 11:39:20 GMT: INFO (sindex): (populate.c:328) {test} populating sindex tokens_key_index ...
Aug 27 2024 11:39:20 GMT: WARNING (particle): (cdt.c:4888) invalid msgpack: has_nonstorage 1 b 0xfffebfe25022 sz 3 i 0
Aug 27 2024 11:39:20 GMT: CRITICAL (particle): (particle_map.c:994) unexpected
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:259) SIGUSR1 received, aborting Aerospike Community Edition build 7.0.0.0 os debian12 arch aarch64 sha 12c9273
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:293) si_code SI_TKILL (-6)
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:386) stacktrace: registers: fault 0000000000000000 x0 0000000000000000 x1 0000000000000000 x2 0000000000000000 x3 0000000000000000 x4 0000000000000000 x5 0000000000000000 x6 0000000000000000 x7 0000000000000000 x8 0000000000000000 x9 0000000000000000 x10 0000000000000000 x11 0000000000000000 x12 0000000000000000 x13 0000000000000000 x14 0000000000000000 x15 0000000000000000 x16 0000000000000000 x17 0000000000000000 x18 0000000000000000 x19 0000fffe7ecc91b8 x20 0000fffe7ecc91d0 x21 0000fffe7ecc9038 x22 0000000000000000 x23 0000fffe7ecc9020 x24 0000fffe7ecc8ff8 x25 0000000000000001 x26 0000000000000061 x27 0000000000000060 x28 0000fffe7d012000 x29 0000fffe7ecc8f30 x30 0000aaaae58bfda0 x31 0000fffe7ecc8f30 pc 0000aaaae58bfda0
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:398) stacktrace: found 18 frames: 0xf7170 0xf7570 0xffffac3a07a0 0xffffab8d09d0 0xffffab88a76c 0x19fe84 0xe4878 0x942f8 0x95804 0x158be4 0x1596a4 0x156b78 0xc10b4 0xc14fc 0x156e60 0x190b64 0xffffab8cee18 0xffffab937e9c offset 0xaaaae5720000
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 0: asd(log_stack_trace base/signal.c:389) [0xf7170]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 1: asd(sig_trace_and_reraise base/signal.c:183) [0xf7570]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 2: linux-vdso.so.1() [0xffffac3a07a0]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 3: /lib/aarch64-linux-gnu/libc.so.6() [0xffffab8d09d0]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 4: /lib/aarch64-linux-gnu/libc.so.6(raise) [0xffffab88a76c]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 5: asd(cf_log_write_no_return /work/source/cf/src/log.c:610) [0x19fe84]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 6: asd(base/particle_map.c:994) [0xe4878]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 7: asd(cdt_context_dig base/cdt.c:2705) [0x942f8]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 8: asd(bin_cdt_get_by_context_vec base/cdt.c:1698 as_bin_cdt_get_by_context base/cdt.c:3062) [0x95804]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 9: asd(sbin_from_bin sindex/sindex.c:1204) [0x158be4]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 10: asd(as_sindex_put_rd sindex/sindex.c:291) [0x1596a4]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 11: asd(populate_reduce_cb sindex/populate.c:453) [0x156b78]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 12: asd(as_index_sprig_reduce base/index.c:459) [0xc10b4]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 13: asd(as_index_reduce_from base/index.c:262) [0xc14fc]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 14: asd(run_populate sindex/populate.c:395) [0x156e60]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 15: asd(joinable_shim_fn /work/source/cf/src/cf_thread.c:421) [0x190b64]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 16: /lib/aarch64-linux-gnu/libc.so.6() [0xffffab8cee18]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:405) stacktrace: frame 17: /lib/aarch64-linux-gnu/libc.so.6() [0xffffab937e9c]
Aug 27 2024 11:39:20 GMT: WARNING (as): (signal.c:473) stacktrace: memory x19 0xfffe7ecc9170 70 94 dc 7e fe ff 00 00 e4 2b 89 e5 aa aa 00 00 a0 00 00 00 00 00 00 00 a0 a4 44 ab 02 00 00 00 22 00 00 00 00 00 00 00 20 50 e2 bf fe ff 00 00 09 00 00 00 05 00 00 00 20 50 e2 bf fe ff 00 00 09 00 00 00 09 00 00 00 01 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 98 91 cc 7e fe ff 00 00 20 94 cc 7e fe ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

**Comment by khaf** [comment]
@kportertx FYI.

**Comment by xorphox** [comment]
Was already fixed in 7.0.0.8.
Please upgrade to at least the latest in the 7.0.x line, which is currently 7.0.0.14.
Consider upgrading to the latest version (currently 7.1.x) for more fixes and improvements.
